### DISCUSSION TITLE
OpenAPI as source of truth

### DISCUSSION THREAD

**Discussion by kdisneur** [cuelang/cue#544] [General]
Hello
 
### context

The source of truth in our company is JSON-API. It’s what we agree on when we implement something (endpoint (verb + path) but also input fields,  data validation, description/explanation and possible errors)
When we all agree on this, we start by generating structs from the JSON-Schema part of the endpoint (input / output). We also run the validation based on a JSON-Schema library.
When we need to test our new endpoint, we generate an OpenAPI client from the OpenAPI file and test from it.

It has good advantages as we guarantee the doc is not an afterthought, we really think about what we implement/expose, we also better detect breaking changes when we do update (e.g if a PR changes some Input/output of existing tests done from the OpenAPI client).

Unfortunately the current process is far from perfect because:
- the packages able to generate go struct from JSON-Schema doesn’t include validations so we need to use two packages one from generating the Go struct at compile time and one that evaluates the JSON-Schema file at runtime to run validations.
- Open-API doesn’t support very well JSON-Schema in the current version so we can’t just import a JSON-Schema as an Open-API payload input/output. It should be fixed with OpenAPI 3.1 but it’s not here yet.

### question

Going back to clue, I understand it would be less powerful than using clue directly but I’m wondering if it was possible to generate cue files and Go structs starting from OpenAPI. I can only see examples the other way around.

Thanks!

**Comment by mpvl** [comment]
It is possible to generate CUE from JSON Schema, but OpenAPI to CUE or CUE to Go isn't there yet. 

In general, the individual translation adaptors have been implemented on a need-to-have basis.

The JSON schema to CUE infrastructure could be reused to also support OpenAPI to CUE.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/544.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
